Feature/authority registry

ejfrancis opened the pull request:
Alright this is a big change. ProseMeteor now automatically scales across any number of application servers.
- Now when a client creates a new editor instance, and calls the Meteor Method `ProseMeteor.openDoc`, it will be given the full URL of the server hosting the doc. All subsequent client communcation for that doc is done via a new DDP connection to that particular server
- each entry in the authority registry db collection maintains a list of all docIds and groupIds currently being hosted on that authority server (these are never cleared out at the moment, that will be addressed in [Issue 18](https://github.com/prosemeteor/prosemirror/issues/18))
- `authority-manager.js` `openDoc()`  will now choose the right authority to host the doc (or choose the authority already hosting either the docId or the doc's groupId). It defaults to preferring the server hosting the least number of documents currently
- Health checks are run against servers before they are chosen to make sure they're up (since we can't be sure they will be). Health checks are also run on a set interval (30 seconds currently), and if any server fails too many consecutively (5 currently) it's assumed to be down and removed from the registry. If it fails 4 but then passes the next one, it's failed count is reset to 0 and it's assumed to be up and healthy.
- Added a `reset-demo` shell script similar to `run-demo` for development and testing purposes

If you pull down this branch and run it, it should all just work as it was previously, but you can watch the logs to see it running all of the logic to select authority servers and run health checks against them.

for #8 and #20 

@funkyeah @ellery44 @archonic

ejfrancis committed:
setting up collection and ValidatedMethods for registering Authorities and adding/removing docs from Authorities in the registry

ejfrancis committed:
added npm ip package and doc schema example

ejfrancis committed:
grab ip from package instead of pass in for register/unregister authority, call methods

ejfrancis committed:
make authority-manager store ip and port on startup and register itself

ejfrancis committed:
cleaned up error aliases and fixed Meteor.Errors when not in method to just log error for now. passed on port to register authority

ejfrancis committed:
dont pass port to unregister

ejfrancis committed:
on startup if server is already registered as authority, clear it out and reinsert as empty

ejfrancis committed:
pass port to unregister and isRegistered since multiple apps could run on same ip

ejfrancis committed:
setting up authority-registry to run health checks when registering a new authority server

ejfrancis committed:
Added runHealthChecks() in AuthorityRegistry to run health checks against each server and remove servers that failed over a threshold

ejfrancis committed:
reset failed health check count on servers that pass but failed previously

ejfrancis committed:
added check for args, dont setup client methods until server is registered

ejfrancis committed:
moved ip and port to be stored in AuthorityRegistry

ejfrancis committed:
removed eslint rule to require backticks, added startHealthCheckInterval() to authority registry to continuously run health checks, added geTUrlOfAuthorityHostingDoc() and accompanying chooseAuthoritytoHostDoc() which takes groupIds into account and selects server with lowest number of docs

ejfrancis committed:
docblocks for authority

ejfrancis committed:
use cb instead of next like other code

ejfrancis committed:
added registry runSingleHealthCheck(), made runHealthChecks() accept array of server objects, made collab-client create and store ddp connection of propery authority, made authority-manager openDoc() return ip of chosen server, added fullUrl property to authorities in db

ejfrancis committed:
added StreamerManager class to manage streamer instances in collab-editor

ejfrancis committed:
fix this to self

ejfrancis committed:
added reset_demo script

ejfrancis committed:
implemented registry.addDocToAuthority(), changed authorityManager.createAuthority() add doc to authority with docId and groupId. fixed groupId behavior in openDoc(). made registry return array of eligible servers

ejfrancis committed:
cleanup logic of selecting an authority to host a doc

ejfrancis committed:
fixed manager/registry logic for handling groupId when group exists on local server, moved instructServerToCreateAuthority() to separate function

eaddingtonwhite commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+    this.port = parseInt(process.env.PORT);
+
+    const ipAndPort = `${this.ip}:${this.port}`;
+    this.fullUrl = (this.port === HTTPS_PORT) ? `https://${ipAndPort}` : `http://${ipAndPort}`;
```
I think here it would maybe be better to specify connection protocol in separate env variable instead of trying to be clever about it and switch depending on port user specified.

ejfrancis replied:
alright now protocol is specified in a `ProseMeteorServer` constructor arg and set in a new singleton ProseMeteorConfig class. new commits coming soon

eaddingtonwhite commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  */
+  setUpHealthCheckEndpoint () {
+    // create a health check endpoint, so other servers can verify this server is alive or not
+    WebApp.rawConnectHandlers.use('/prosemeteor-health-check', (req, res, cb) => {
```
Might be cool to make this configurable.

eaddingtonwhite commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  * @param {Function} cb          callback, with number of failed connections passed to it
+  * @returns {Number} number of failed connections
+  */
+  runHealthChecks (serversArray) {
```
Maybe break this function up. Its pretty long and there is a lot going on.

Something like this:
CheckServers()
StoreResults()
updateHealthRegistry()

ejfrancis replied:
moved health check logic to its own module that exports functions for use in `ProseMeteorAuthorityRegistry`, commits incoming

eaddingtonwhite commented on lib/imports/api/authority/server/authority-manager.js:
```diff
@@ -58,47 +75,151 @@ export class AuthorityManager {
 
   /**
-  * Open a ProseMirror document. Creates a new Authority to track the doc.
+  * Open a ProseMirror document. Instructs appropriate server to create a server in memory and returns the server's URL.
```
I would try to think of some other name for Authority then server. This can get confusing. ex: in this comment "find server to create server on". Maybe im reading this wrong and comment should just be cleaned up.

ejfrancis replied:
oh yeah that's confusing. It should read `Instructs appropriate server to create an Authority in memory and returns the server's URL`.

eaddingtonwhite commented on lib/imports/api/authority-registry/server/methods.js:
```diff
@@ -0,0 +1,145 @@
+ recentFailedConnectionTimes: [13662212879613, 13662212879721, 1366221289532]
+}
+*/
+export const insertAuthority = new ValidatedMethod({
```
You should have method docs on all exported functions for this module.

ejfrancis replied:
I've added documentaiton to all methods and changed names for better clarity of authority vs authority server, coming soon

eaddingtonwhite commented:
First pass looking good! 

Few first thoughts:
- Start using info and debug logging. This will make code more readable and also help turn off noise when monitoring production system. I would check out https://www.npmjs.com/package/loglevel
- I would use a UUID for tracking nodes instead of their IP's and Port's This will make managing them easier and more future proof especially as this scales up to 100's of nodes. 
- I would add fast authority node migration and conflict resolution as a necessary part of this feature. This will help you do things like deploy newer code without effecting users on the system. For instance how would you move an authority session off of this node quickly without a user noticing. This will help make your deploys faster and safer. Also by building code this way it will make error scenarios easier to handle. For instance what do you do when you loose the server hosting 1000's of authorities. Would there be any data loss? Its hard to tell just looking at this PR but how would local changes be handled and how would you handle loosing latest doc state in an authority. Also what server is Authority manager running on or is it just general interface with DB?  
- I would start adding some unit tests. This project is starting to get pretty big for not having any. 

Thats all for now ill keep looking over and add more later. Looks good.

ejfrancis commented:
> Start using info and debug logging. 

yeah for logging, we'll need an issue to implement a true logger with info/warning/error/debug logs. right now there's just `console.log`s scattered everywhere for debugging purposes. I've just added #22

> I would use a UUID for tracking nodes instead of their IP's and Port's 

Each server already has one in the registry, so that shouldn't be hard. Not a bad idea, although technically the ip + port _should_ always be unique and is a bit more human friendly to read.

>  For instance what do you do when you loose the server hosting 1000's of authorities. Would there be any data loss? Its hard to tell just looking at this PR but how would local changes be handled and how would you handle loosing latest doc state in an authority. Also what server is Authority manager running on or is it just general interface with DB?

As of right now we have no logic to remove an in-memory session of a document. That will be part of #18. Document's have snapshots stored to db on an interval so currently if a server went down, whatever the latest snapshot was would be what the document had next time it's opened. I like the idea of migrating sessions, but I think that'll add a lot of scope to this PR, so I'm going to add another issue around that and implement it separately. If a server just shuts off without warning, it would be hard to migrate the existing data in memory to another server, but if it's given a `SIGINT`/`SIGTERM`/`SIGKILL` due to a rolling deployment, we could capture that and migrate sessions gracefully. I just added #23 for that work, should be interesting.

> I would start adding some unit tests. This project is starting to get pretty big for not having any.

Yeah I've been hesitant to write any so far because the API is changing so rapidly and the overall design isn't 100% locked in stone yet. I'm not a fan of writing tests and then immediately invalidating them with a breaking API change, just seems like unnecessary work. But with this PR in, I think at least the server-side authority logic will be mostly the same for now on, so I can start adding unit tests. The client will be changing pretty heavily when it's moved to a ProseMirror plugin, so I'm hesitant to test that code since tests would break really soon after writing and right now we don't have customers to worry about.

ejfrancis commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+    this.port = parseInt(process.env.PORT);
+
+    const ipAndPort = `${this.ip}:${this.port}`;
+    this.fullUrl = (this.port === HTTPS_PORT) ? `https://${ipAndPort}` : `http://${ipAndPort}`;
```
yeah I like that idea. when a user creates a `new ProseMeteorServer()` they can have the option to pass in the protocol, and we could fallback to trying to figure it out like this if they don't provide it.

eaddingtonwhite replied:
I would not fall back at all like this. It just adds complexity for no gain really. I would copy how AWS does health checks on ELB's for reference. For example what if I want TCP health check on port 443?

https://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/elb-healthchecks.html

ejfrancis replied:
yeah thinking more about this I think I'll just let the application developer specify their protocol :+1:

gckellerman commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  *  Starts a continuous interval to run health checks against all other servers so
+  *  we can know when an authority server is no longer available.
+  */
+  startHealthCheckInterval () {
```
im having a little trouble understanding the intended  state diagram for the health checks... especially the fail paths...

if a previously available server is no longer available you retry a definable number of times and then remove it.... is the intention to automatically re-open it on some other server

unregistering looks like its a hard command but once docs can be removed from an authority should unregistering be a longer process where the server no longer accepts new docs and waits until its last doc has been closed to truly unregister... or do we migrate the docs on the server to be unregistered to a new server

ejfrancis replied:
So actually `unregister()` is never being used at the moment. The only time I can imagine that would be used would be if the server received a `SIGINT`/`SIGTERM`/`SIGKILL` to kill it, such as for a rolling deployment. that would be covered in [the issue for migrating sessions](https://github.com/prosemeteor/prosemirror/issues/23). But yes, I imagine in that case we'd want to gracefully migrate each existing doc to a new server, wait for all users to disconnect from this server, unregister this authority server from the registry, and then allow shutdown.

Right now the only time we remove an authority server from the registry is when an authority server has failed 5 consecutive health checks, since we assume that server is no longer up.

ejfrancis replied:
> if a previously available server is no longer available you retry a definable number of times and then remove it.... is the intention to automatically re-open it on some other server

The intention is just to remove that server from the registry so it stops being chosen as an eligible server to host authorities, since it's no longer alive. If it's failed 5 consecutive health checks, there are obviously no more documents in memory to migrate since the server has been offline for a bit. Normally you'd do load balancing with something like nginx and you don't need to mess with this stuff at the application level, but with Meteor I'm going for the package philosophy that a package should "just work" if it's dropped into an application.

gckellerman commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  * @param {String} [params.groupId]  optional, a groupId of this doc to associate it with other docs
+  */
+  addDocToAuthority ({ docId, groupId = null }) {
+    addDocToAuthority.call({ docId, groupId, fullUrl: this.fullUrl });
```
grouping docs with others like docs seems like a no-brainer at first thought but what does it solve? is the idea just that group docs will all have similar latency? - if so, does that imply that the first doc in a group has to be added by the lowest latency server available (which isn't really the case since it goes to the server that is manage the least amount of docs.)

ejfrancis replied:
prosemirror doesn't really have a concept of "pages", it just has full documents, so I'm not sure that's possible in its current stage. perhaps that's something that will be added later on, the proejct is still early in development and changing rapidly. regardless of how large a document is though, you're only sending and receiving operations of a couple keystrokes at a time so the size of a document shouldn't affect scaling too much. we'll just have to limit the size of operations history for each doc kept in memory on the server, and allow the package user to specify that limit. I agree with the concerns of having groups possibly affect scaling depending on the use case of the package user, so I'm inclined to say that we provide a config option that lets you enable/disable it

gckellerman commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+      })
+      .slice(0, 5);
+    console.log(`ProseMeteor AuthorityRegistry: Selecting eligible authority servers with least documents: "${JSON.stringify(fullRegistrySortedByDocIdsCount, null, 2)}"`);
+
```
Im assuming with a single user the server load is dominated just by opening and managing the documnent, but at some number of users having the same doc open the server load of managing those users becomes greater... thoughts on where this crossover point might be - its probably also a memory/processing tradeoff too I suppose

eaddingtonwhite replied:
It would be interesting to figure out where bottle neck would be with real users. Ideally what ever we go with should be very easy to scale at a constant rate of node per # of docs open.

ejfrancis replied:
This is really hard to answer without testing it on a few servers under load, and even harder because it would depend on the application user's implementation. Right now all of the db load is when a user connects to a document to query the db to see if any other servers are hosting the doc, or to delegate a document to a server. Once users are connected to a server, it's all just reading and writing from memory for every keystroke so that won't be much db load and JavaScript is pretty damn fast for stuff like that. Memory limitations will come into play with a huge doc in memory, but we can limit the size of the step history we keep in memory on the server.

I think at the end of the day, we can't predict how this package will be used and what the right scaling decisions will be that work for everyone, so our best choice is to allow the application developer to specify their own options for things like max number of users who can edit a doc, snapshot interval time, max history size, etc, with some intelligent defaults that we provide that would work for basic usage.

eaddingtonwhite replied:
I agree I think above discussion about how to distribute pages well is much more important at this stage of this project. In future as system scales up we would probably break authentication system out into a micro service and it would be irrelevant.

eaddingtonwhite commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  * @param {String} [params.groupId]  optional, a groupId of this doc to associate it with other docs
+  */
+  addDocToAuthority ({ docId, groupId = null }) {
+    addDocToAuthority.call({ docId, groupId, fullUrl: this.fullUrl });
```
I agree with this. Unless there is very good reason for this it could lead to scaling problems.

ejfrancis replied:
The thought there is that if I'm viewing a page with 15 editable docs (which will be valid in the application I'm planning on using this in), modern web browsers have a [hard limit on number of connections per domain](http://sgdev-blog.blogspot.com/2014/01/maximum-concurrent-connection-to-same.html) so you could group related documents (i.e. viewed on the same page) all together so they're only on one. Perhaps since we're connecting directly via IP, this wouldn't be an issue since it's a limit per hostname? If anyone knows more about this I'm open to discussion, but the connection limit was my main motivation behind the grouping decision. I do agree that having a group of 500 documents on one server could bring it to a standstill and affect all users connected to that server, so perhaps we need more limiting around number allowed per group, or the option to distribute groups > some number to other servers.

eaddingtonwhite replied:
I would do a 1 page per node and distribute pages randomly across all nodes. Then on client side preload previous and next pages authority so editing is seamless to user. This would mean clients only have 3 connections open at any given moment as well allowing system to scale evenly regardless of how large the documents being edited are. 

I would still keep groupId in schema though for logging and future proofing system though.

ejfrancis commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -0,0 +1,319 @@
+  * @param {Function} cb          callback, with number of failed connections passed to it
+  * @returns {Number} number of failed connections
+  */
+  runHealthChecks (serversArray) {
```
:+1: will do

ejfrancis committed:
created singleton ProseMeteorConfig class to store global config values and allow setting/getting, added documentation to registry methods, refactored health check logic into its own module

ejfrancis commented:
Alright @ellery44  and @funkyeah I've addressed the comments above
- moved health check logic into its own module `health-checks.js` and refactored it into smaller methods
- created a singleton of a new class `ProseMeteorConfig` for getting/setting global config values that's used throughout the server code
- allow all config values on server to be passed into `ProseMeteorServer` constructor
- removed logic to try and guess the protocol (http or https), instead allowed it to be passed in as a config value
- added documentation to `ValidatedMethods`s for authority registry
- renamed `ValidatedMethod`s for authority registry to make the difference between "authority" and "authority server" more clear. "authority" is a single class maintaining state of a document in memory when its open, "authority server" is a server that's in the registry in db and is capable of hosting ProseMeteor authorities

The last piece will be removing document authorities from memory when no more users are editing the doc. That will be done in #18 . And eventually I think we'll need to stop using a global streamer called 'prosemirror-pipe' in favor of individual streamers for each doc, i.e. 'prosemirror-pipe-someDocId', that way we can authenticate on a per-streamer basis.

ejfrancis commented on demo/server/main.js:
```diff
@@ -6,4 +6,7 @@ Meteor.startup(() => {
 
 // setup ProseMeteor on server
-let proseMeteorServer = new ProseMeteorServer({ snapshotIntervalMs: 5000 });
+let proseMeteorServer = new ProseMeteorServer({
```
config values are passed into `ProseMeteorServer` so the user of the package can override defaults

ejfrancis commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -1,35 +1,34 @@
-  removeAuthorities,
-  resetFailedHealthChecksCount,
-  addDocToAuthority
+  insertAuthorityServerIntoRegistry,
```
renamed these methods for better clarity

ejfrancis commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -1,35 +1,34 @@
-const HEALTH_CHECK_INTERVAL_MS = 1000 * 30;   // interval on which to run health checks
+import { proseMeteorConfig } from './../../config/server/prosemeteor-config';
+import {
+  checkHealthOfServers,
```
moved health check logic out into its own module with smaller methods

ejfrancis commented on lib/imports/api/authority-registry/server/authority-registry.js:
```diff
@@ -1,35 +1,34 @@
   constructor () {
-    this.ip = ip.address();
-    this.port = parseInt(process.env.PORT);
+    proseMeteorConfig.ip = ip.address();
```
storing global config like this server's ip and port in the single config class instance

ejfrancis commented on lib/imports/api/authority-registry/server/methods.js:
```diff
@@ -13,22 +12,33 @@ Authority schema looks like this
-    const fullUrl = `${ip}:${port}`;
-    const fullUrlWithProtocol = (this.port === HTTPS_PORT) ? `https://${fullUrl}` : `http://${fullUrl}`;
+    const ipAndPort = `${ip}:${port}`;
+    const fullUrl = `${proseMeteorConfig.protocol}://${ipAndPort}`;
```
got rid of trying to check the port, instead getting it from the config

ejfrancis commented on lib/imports/api/authority-registry/server/methods.js:
```diff
@@ -88,26 +113,39 @@ export const addDocToAuthority = new ValidatedMethod({
-export const removeDocFromAuthority = new ValidatedMethod({
-  name: 'authorityRegistry.removeDocFromAuthority',
+/**
+* Removes a document from an authority server in the registry.
```
added documentation for all these methods

ejfrancis commented on lib/imports/api/config/server/prosemeteor-config.js:
```diff
@@ -0,0 +1,62 @@
+/**
+* Configuration class for getting and setting config values for ProseMeteor.
+*/
+class ProseMeteorConfig {
```
new config class with getters/setters for all config values. the module exports a single instance, making it a singleton

ejfrancis commented on lib/imports/api/prosemeteor/server/prosemeteor-server.js:
```diff
@@ -1,31 +1,46 @@
   */
-  constructor ({ snapshotIntervalMs = 5000 }) {
+  constructor ({
+    // set default config values if not provided.
```
all default config values are defined here in the `ProseMeteorServer` constructor

ejfrancis committed:
remove redundant check

ejfrancis commented:
can someone review new changes? I think I addressed all comments. I'll tackle testing after this is merged, once our API is a little more stable. Scaling issues/tuning comments are all valid but will have to be faced once we can properly test this under load, this PR was more just to prove that this can be run across multiple servers with manually configuring all available servers.

funkyeah commented:
I think it looks pretty damn good and should be merged. I walked through most of the code again and it feels pretty well laid out. 

Took some notes along the way to make sure I'm understanding everything correctly. The bolded things are items that stuck out to me but certainly not anything that needs to be addressed.

demo app server creates new instance of ProseMeteorServer on every instance the app is scaled across
- demo does not pass all config options
  - **missing health check interval**
  - **missing failed attempts threshold**

ProseMeteorServer
- grabs document collection
- creates streamer pipe
  - passes both to new instance of Authority Manager

Authority Manager
- creates new AuthorityRegistry
- **has some methods that seem more related to Authority than Authority Manager (latestDocState, clientSubmitSteps, stepsSince etc...)**
- wouldn't these make sense to be done against the Authority that manages the doc... or are we saying these are conviniences for getting the doc faster / without connecting to the authority

AuthorityRegistry
- **calls this.setupClientMethods but that function is empty**

demo app calls openDoc of AuthorityManager
- checks if current server is managing doc
- checks if other server is managing doc
- creates an Authority on the next eligible server if neither of the above and adds it to the registry

Authority
- handles document updates

ejfrancis commented:
> demo does not pass all config options

Yeah right now those that aren't passed in will use [the default values](https://github.com/ejfrancis/prosemirror/blob/be3815b8a135d179d78def3579c7c792b7892683/lib/imports/api/prosemeteor/server/prosemeteor-server.js#L18), so the demo's just using those.

> AuthorityRegistry calls this.setupClientMethods but that function is empty

This was leftover from earlier testing. It's not being used atm, but I just left it there because I didn't know if I'd need to add some methods in there at one point. It can be ripped out safely right now.

> AuthorityManager has some methods that seem more related to Authority than Authority Manager (latestDocState, clientSubmitSteps, stepsSince etc...)

Yeah those methods are really directly for individual Authority instances. But when we define those Meteor Methods we need access to the state of every Authority, so we have to declare them in the AuthorityManager. It wouldn't be possible to have each separate Authority declare it's own methods since they'd all have the same names (i.e. "stepsSince") and they'd override each other, so I'm defining the Method in the place where we can get access to the state of all Authority instances, which is the AuthorityManager.

ejfrancis commented:
Alright it's merged! Next I'll work on the logic of detecting when clients disconnect and removing Authority instances from memory when there are no more clients. After that we'll be in a pretty good spot to work on authentication

ejfrancis commented:
For future reference this is what the current communication looks like. Note that #5 is not implemented yet and will be done in #18 

![prosemeteor_architecture](https://cloud.githubusercontent.com/assets/3171352/18620003/c17f05a4-7dbe-11e6-9232-79c75191cae2.png)

@ellery44
